From 37cd7d2807046dbac8e233d9387adee24fa8ce74 Mon Sep 17 00:00:00 2001
From: Lionel Debieve <lionel.debieve@foss.st.com>
Date: Wed, 8 Dec 2021 17:35:24 +0100
Subject: [PATCH 0333/1695] drivers: stm32_exti: add stm32mp_exti

Add support of EXTI (Extended Interrupt and event) controller. This
hardware block is used mainly to wake up platform from low power mode
(except STANDBY mode) from a peripheral connected to an EXTI line
(i.e. USB, ETH(WOL), CEC ...) and to handle GPIO interrupt.

This driver offers several APIs to configure EXTI lines and it targets
only STM32MP13x SoCs.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 849427b2d9c740cb8b4ea7fa328f78177f960b9e drivers: stm32_exti: add stm32mp_exti]
Change-Id: Ia8e1348b4563ea5f43e997a49145de5706f1c72b
Signed-off-by: Alexandre Torgue <alexandre.torgue@foss.st.com>
---
 core/drivers/stm32_exti.c         | 375 ++++++++++++++++++++++++++++++
 core/drivers/sub.mk               |   1 +
 core/include/drivers/stm32_exti.h |  73 ++++++
 3 files changed, 449 insertions(+)
 create mode 100644 core/drivers/stm32_exti.c
 create mode 100644 core/include/drivers/stm32_exti.h

diff --git a/core/drivers/stm32_exti.c b/core/drivers/stm32_exti.c
new file mode 100644
index 000000000..c75246167
--- /dev/null
+++ b/core/drivers/stm32_exti.c
@@ -0,0 +1,375 @@
+// SPDX-License-Identifier: BSD-3-Clause
+/*
+ * Copyright (c) 2021, STMicroelectronics
+ */
+
+#include <drivers/stm32_exti.h>
+#include <io.h>
+#include <kernel/boot.h>
+#include <kernel/dt.h>
+#include <kernel/dt_driver.h>
+#include <kernel/pm.h>
+#include <kernel/spinlock.h>
+#include <libfdt.h>
+#include <mm/core_memprot.h>
+#include <tee_api_types.h>
+
+/* Registers */
+#define _EXTI_CR1	0x60U
+#define _EXTI_CR2	0x64U
+#define _EXTI_CR3	0x68U
+#define _EXTI_CR4	0x6CU
+
+struct stm32_exti_bank {
+	uint32_t imr_ofst;
+	uint32_t rtsr_ofst;
+	uint32_t ftsr_ofst;
+	uint32_t rpr_ofst;
+	uint32_t fpr_ofst;
+	uint32_t tzenr_ofst;
+	uint32_t wakeup;
+	uint32_t imr_cache;
+	uint32_t rtsr_bkp;
+	uint32_t ftsr_bkp;
+	uint32_t tz_bkp;
+};
+
+static uint32_t port_sel_bkp[4];
+
+struct stm32_exti_priv {
+	vaddr_t base;
+};
+
+static struct stm32_exti_priv stm32_exti;
+
+static struct stm32_exti_bank stm32mp1_exti_b1 = {
+	.imr_ofst	= 0x80,
+	.rtsr_ofst	= 0x00,
+	.ftsr_ofst	= 0x04,
+	.rpr_ofst	= 0x0C,
+	.fpr_ofst	= 0x10,
+	.tzenr_ofst	= 0x14,
+};
+
+static struct stm32_exti_bank stm32mp1_exti_b2 = {
+	.imr_ofst	= 0x90,
+	.rtsr_ofst	= 0x20,
+	.ftsr_ofst	= 0x24,
+	.rpr_ofst	= 0x2C,
+	.fpr_ofst	= 0x30,
+	.tzenr_ofst	= 0x34,
+};
+
+static struct stm32_exti_bank stm32mp1_exti_b3 = {
+	.imr_ofst	= 0xA0,
+	.rtsr_ofst	= 0x40,
+	.ftsr_ofst	= 0x44,
+	.rpr_ofst	= 0x4C,
+	.fpr_ofst	= 0x50,
+	.tzenr_ofst	= 0x54,
+};
+
+static struct stm32_exti_bank *stm32mp1_exti_banks[] = {
+	&stm32mp1_exti_b1,
+	&stm32mp1_exti_b2,
+	&stm32mp1_exti_b3,
+};
+
+/* EXTI access protection */
+static unsigned int lock = SPINLOCK_UNLOCK;
+
+static uint32_t stm32_exti_get_bank(uint32_t exti_line)
+{
+	uint32_t bank = 0;
+
+	if (exti_line <= 31)
+		bank = 0;
+	else if (exti_line <= 63)
+		bank = 1;
+	else if (exti_line <= 95)
+		bank = 2;
+	else
+		panic();
+
+	return bank;
+}
+
+void stm32_exti_set_type(uint32_t exti_line, uint32_t type)
+{
+	uint32_t mask = BIT(exti_line % 32);
+	uint32_t r_trig = 0;
+	uint32_t f_trig = 0;
+	uint32_t exceptions = 0;
+	unsigned int i = 0;
+
+	switch (type) {
+	case EXTI_TYPE_RISING:
+		r_trig |= mask;
+		f_trig &= ~mask;
+		break;
+	case EXTI_TYPE_FALLING:
+		r_trig &= ~mask;
+		f_trig |= mask;
+		break;
+	case EXTI_TYPE_BOTH:
+		r_trig |= mask;
+		f_trig |= mask;
+		break;
+	default:
+		panic();
+	}
+
+	i = stm32_exti_get_bank(exti_line);
+
+	exceptions = cpu_spin_lock_xsave(&lock);
+
+	io_mask32(stm32_exti.base + stm32mp1_exti_banks[i]->rtsr_ofst, r_trig,
+		  mask);
+	io_mask32(stm32_exti.base + stm32mp1_exti_banks[i]->ftsr_ofst, f_trig,
+		  mask);
+
+	cpu_spin_unlock_xrestore(&lock, exceptions);
+}
+
+void stm32_exti_mask(uint32_t exti_line)
+{
+	uint32_t val = 0;
+	uint32_t mask = BIT(exti_line % 32);
+	uint32_t exceptions = 0;
+	unsigned int i = 0;
+
+	i = stm32_exti_get_bank(exti_line);
+
+	exceptions = cpu_spin_lock_xsave(&lock);
+
+	val = io_read32(stm32_exti.base + stm32mp1_exti_banks[i]->imr_ofst);
+	val &= ~mask;
+	io_write32(stm32_exti.base + stm32mp1_exti_banks[i]->imr_ofst, val);
+	stm32mp1_exti_banks[i]->imr_cache = val;
+
+	cpu_spin_unlock_xrestore(&lock, exceptions);
+}
+
+void stm32_exti_unmask(uint32_t exti_line)
+{
+	uint32_t val = 0;
+	uint32_t mask = BIT(exti_line % 32);
+	uint32_t exceptions = 0;
+	unsigned int i = 0;
+
+	i = stm32_exti_get_bank(exti_line);
+
+	exceptions = cpu_spin_lock_xsave(&lock);
+
+	val = io_read32(stm32_exti.base + stm32mp1_exti_banks[i]->imr_ofst);
+	val |= mask;
+	io_write32(stm32_exti.base + stm32mp1_exti_banks[i]->imr_ofst, val);
+	stm32mp1_exti_banks[i]->imr_cache = val;
+
+	cpu_spin_unlock_xrestore(&lock, exceptions);
+}
+
+void stm32_exti_enable_wake(uint32_t exti_line)
+{
+	uint32_t mask = BIT(exti_line % 32);
+	uint32_t exceptions = 0;
+	unsigned int i = 0;
+
+	i = stm32_exti_get_bank(exti_line);
+
+	exceptions = cpu_spin_lock_xsave(&lock);
+
+	stm32mp1_exti_banks[i]->wakeup |= mask;
+
+	cpu_spin_unlock_xrestore(&lock, exceptions);
+}
+
+void stm32_exti_disable_wake(uint32_t exti_line)
+{
+	uint32_t mask = BIT(exti_line % 32);
+	uint32_t exceptions = 0;
+	unsigned int i = 0;
+
+	i = stm32_exti_get_bank(exti_line);
+
+	exceptions = cpu_spin_lock_xsave(&lock);
+
+	stm32mp1_exti_banks[i]->wakeup &= ~mask;
+
+	cpu_spin_unlock_xrestore(&lock, exceptions);
+}
+
+void stm32_exti_clear(uint32_t exti_line)
+{
+	uint32_t mask = BIT(exti_line % 32);
+	uint32_t exceptions = 0;
+	unsigned int i = 0;
+
+	i = stm32_exti_get_bank(exti_line);
+
+	exceptions = cpu_spin_lock_xsave(&lock);
+
+	io_mask32(stm32_exti.base + stm32mp1_exti_banks[i]->rpr_ofst, mask,
+		  mask);
+	io_mask32(stm32_exti.base + stm32mp1_exti_banks[i]->fpr_ofst, mask,
+		  mask);
+
+	cpu_spin_unlock_xrestore(&lock, exceptions);
+}
+
+void stm32_exti_set_tz(uint32_t exti_line)
+{
+	uint32_t mask = BIT(exti_line % 32);
+	uint32_t exceptions = 0;
+	unsigned int i = 0;
+
+	i = stm32_exti_get_bank(exti_line);
+
+	exceptions = cpu_spin_lock_xsave(&lock);
+
+	io_mask32(stm32_exti.base + stm32mp1_exti_banks[i]->tzenr_ofst, mask,
+		  mask);
+
+	cpu_spin_unlock_xrestore(&lock, exceptions);
+}
+
+void stm32_exti_set_gpio_port_sel(uint8_t bank, uint8_t pin)
+{
+	uint32_t reg = _EXTI_CR1 + (pin / 4) * 4;
+	uint32_t shift = (pin % 4) * 8;
+	uint32_t val = bank << shift;
+	uint32_t mask = 0xff << shift;
+	uint32_t exceptions = 0;
+
+	exceptions = cpu_spin_lock_xsave(&lock);
+
+	io_mask32(stm32_exti.base + reg, val, mask);
+
+	cpu_spin_unlock_xrestore(&lock, exceptions);
+}
+
+static void stm32_exti_pm_suspend(void)
+{
+	uint32_t mask = 0;
+	uint32_t base_cr = 0;
+	uint32_t i = 0;
+
+	/* Save ftsr, rtsr and tzen registers */
+	stm32mp1_exti_b1.ftsr_bkp = io_read32(stm32_exti.base +
+					      stm32mp1_exti_b1.ftsr_ofst);
+	stm32mp1_exti_b1.rtsr_bkp = io_read32(stm32_exti.base +
+					      stm32mp1_exti_b1.rtsr_ofst);
+	stm32mp1_exti_b1.tz_bkp = io_read32(stm32_exti.base +
+					    stm32mp1_exti_b1.tzenr_ofst);
+
+	stm32mp1_exti_b2.ftsr_bkp = io_read32(stm32_exti.base +
+					      stm32mp1_exti_b2.ftsr_ofst);
+	stm32mp1_exti_b2.rtsr_bkp = io_read32(stm32_exti.base +
+					      stm32mp1_exti_b2.rtsr_ofst);
+	stm32mp1_exti_b2.tz_bkp = io_read32(stm32_exti.base +
+					    stm32mp1_exti_b2.tzenr_ofst);
+
+	stm32mp1_exti_b3.ftsr_bkp = io_read32(stm32_exti.base +
+					      stm32mp1_exti_b3.ftsr_ofst);
+	stm32mp1_exti_b3.rtsr_bkp = io_read32(stm32_exti.base +
+					      stm32mp1_exti_b3.rtsr_ofst);
+	stm32mp1_exti_b3.tz_bkp = io_read32(stm32_exti.base +
+					    stm32mp1_exti_b3.tzenr_ofst);
+
+	/* Let enabled only the wakeup sources set*/
+	mask = stm32mp1_exti_b1.wakeup;
+	io_mask32(stm32_exti.base + stm32mp1_exti_b1.imr_ofst, mask, mask);
+	mask = stm32mp1_exti_b2.wakeup;
+	io_mask32(stm32_exti.base + stm32mp1_exti_b2.imr_ofst, mask, mask);
+	mask = stm32mp1_exti_b3.wakeup;
+	io_mask32(stm32_exti.base + stm32mp1_exti_b3.imr_ofst, mask, mask);
+
+	/* Save EXTI port selection */
+	for (i = 0; i < 4; i++) {
+		base_cr = stm32_exti.base + _EXTI_CR1;
+		port_sel_bkp[i] = io_read32(base_cr + i * 4);
+	}
+}
+
+static void stm32_exti_pm_resume(void)
+{
+	uint32_t base_cr = 0;
+	uint32_t i = 0;
+
+	/* Restore ftsr and rtsr registers */
+	io_write32(stm32_exti.base + stm32mp1_exti_b1.ftsr_ofst,
+		   stm32mp1_exti_b1.ftsr_bkp);
+	io_write32(stm32_exti.base + stm32mp1_exti_b2.ftsr_ofst,
+		   stm32mp1_exti_b2.ftsr_bkp);
+	io_write32(stm32_exti.base + stm32mp1_exti_b3.ftsr_ofst,
+		   stm32mp1_exti_b3.ftsr_bkp);
+	io_write32(stm32_exti.base + stm32mp1_exti_b1.rtsr_ofst,
+		   stm32mp1_exti_b1.rtsr_bkp);
+	io_write32(stm32_exti.base + stm32mp1_exti_b2.rtsr_ofst,
+		   stm32mp1_exti_b2.rtsr_bkp);
+	io_write32(stm32_exti.base + stm32mp1_exti_b3.rtsr_ofst,
+		   stm32mp1_exti_b3.rtsr_bkp);
+	io_write32(stm32_exti.base + stm32mp1_exti_b1.tzenr_ofst,
+		   stm32mp1_exti_b1.tz_bkp);
+	io_write32(stm32_exti.base + stm32mp1_exti_b2.tzenr_ofst,
+		   stm32mp1_exti_b2.tz_bkp);
+	io_write32(stm32_exti.base + stm32mp1_exti_b3.tzenr_ofst,
+		   stm32mp1_exti_b3.tz_bkp);
+
+	/* Restore imr */
+	io_write32(stm32_exti.base + stm32mp1_exti_b1.imr_ofst,
+		   stm32mp1_exti_b1.imr_cache);
+	io_write32(stm32_exti.base + stm32mp1_exti_b2.imr_ofst,
+		   stm32mp1_exti_b2.imr_cache);
+	io_write32(stm32_exti.base + stm32mp1_exti_b3.imr_ofst,
+		   stm32mp1_exti_b3.imr_cache);
+
+	/* Restore EXTI port selection */
+	for (i = 0; i < 4; i++) {
+		base_cr = stm32_exti.base + _EXTI_CR1;
+		io_write32(base_cr + i * 4, port_sel_bkp[i]);
+	}
+}
+
+static TEE_Result
+stm32_exti_pm(enum pm_op op, unsigned int pm_hint __unused,
+	      const struct pm_callback_handle *pm_handle __unused)
+{
+	if (op == PM_OP_SUSPEND)
+		stm32_exti_pm_suspend();
+	else
+		stm32_exti_pm_resume();
+
+	return TEE_SUCCESS;
+}
+
+#ifdef CFG_EMBED_DTB
+static TEE_Result stm32_exti_probe(const void *fdt, int node,
+				   const void *comp_data __unused)
+{
+	struct dt_node_info dt_info = { };
+	struct io_pa_va base = { };
+
+	fdt_fill_device_info(fdt, &dt_info, node);
+
+	base.pa = dt_info.reg;
+
+	stm32_exti.base = io_pa_or_va_secure(&base, dt_info.reg_size);
+	assert(stm32_exti.base);
+
+	register_pm_core_service_cb(stm32_exti_pm, NULL, "stm32-exti");
+
+	return TEE_SUCCESS;
+}
+
+static const struct dt_device_match stm32_exti_match_table[] = {
+	{ .compatible = "st,stm32mp13-exti" },
+	{ }
+};
+
+DEFINE_DT_DRIVER(stm32_exti_dt_driver) = {
+	.name = "stm32-exti",
+	.match_table = stm32_exti_match_table,
+	.probe = &stm32_exti_probe,
+};
+#endif
diff --git a/core/drivers/sub.mk b/core/drivers/sub.mk
index 8c5219466..1cd68d571 100644
--- a/core/drivers/sub.mk
+++ b/core/drivers/sub.mk
@@ -36,6 +36,7 @@ srcs-$(CFG_AMLOGIC_UART) += amlogic_uart.c
 srcs-$(CFG_MVEBU_UART) += mvebu_uart.c
 srcs-$(CFG_STM32_BSEC) += stm32_bsec.c
 srcs-$(CFG_STM32_BSEC3) += stm32_bsec3.c
+srcs-$(CFG_STM32_EXTI) += stm32_exti.c
 srcs-$(CFG_STM32_GPIO) += stm32_gpio.c
 srcs-$(CFG_STM32_IWDG) += stm32_iwdg.c
 srcs-$(CFG_STM32_I2C) += stm32_i2c.c
diff --git a/core/include/drivers/stm32_exti.h b/core/include/drivers/stm32_exti.h
new file mode 100644
index 000000000..7a758d16c
--- /dev/null
+++ b/core/include/drivers/stm32_exti.h
@@ -0,0 +1,73 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
+/*
+ * Copyright (c) 2021, STMicroelectronics
+ */
+
+#ifndef __STM32_EXTI_H
+#define __STM32_EXTI_H
+
+#include <tee_api_types.h>
+
+#define EXTI_TYPE_RISING	1
+#define EXTI_TYPE_FALLING	2
+#define EXTI_TYPE_BOTH	(EXTI_TYPE_RISING | EXTI_TYPE_FALLING)
+
+/*
+ * Set EXTI type in RTSR and FTSR EXTI registers
+ * @exti_line: EXTI line number
+ * @type: type (rising, falling or both) to set
+ * Return a TEE_Result compliant return value
+ */
+void stm32_exti_set_type(uint32_t exti, uint32_t type);
+
+/*
+ * Mask EXTI Interrupt (IMR)
+ * @exti_line: EXTI line number
+ * Return a TEE_Result compliant return value
+ */
+void stm32_exti_mask(uint32_t exti_line);
+
+/*
+ * Unmask EXTI Interrupt (IMR)
+ * @exti_line: EXTI line number
+ * Return a TEE_Result compliant return value
+ */
+void stm32_exti_unmask(uint32_t exti_line);
+
+/*
+ * Enable EXTI line as wakeup interrupt
+ * @exti_line: EXTI line number
+ * Return a TEE_Result compliant return value
+ */
+void stm32_exti_enable_wake(uint32_t exti_line);
+
+/*
+ * Disable EXTI line as wakeup interrupt
+ * @exti_line: EXTI line number
+ * Return a TEE_Result compliant return value
+ */
+void stm32_exti_disable_wake(uint32_t exti_line);
+
+/*
+ * Clear pending EXTI interrupts
+ * @exti_line: EXTI line number
+ * Return a TEE_Result compliant return value
+ */
+void stm32_exti_clear(uint32_t exti_line);
+
+/*
+ * Configure EXTI mux for GPIO irq
+ * @bank: GPIO bank id
+ * @pin: GPIO number in the bank
+ * Return a TEE_Result compliant return value
+ */
+void stm32_exti_set_gpio_port_sel(uint8_t bank, uint8_t pin);
+
+/*
+ * Securize the EXTI line
+ * @exti_line: EXTI line number
+ * Return a TEE_Result compliant return value
+ */
+void stm32_exti_set_tz(uint32_t exti_line);
+
+#endif /*__STM32_EXTI_H*/
-- 
2.47.2

